caas/kubernetes/provider/cloud.go

```diff
@@ -125,31 +125,44 @@ func UpdateKubeCloudWithStorage(k8sCloud *cloud.Cloud, credential cloud.Credenti
 	// If the user has not specified storage, check that the cluster has Juju's opinionated defaults.
 	cloudType := cloud.SplitHostCloudRegion(storageParams.HostCloudRegion)[0]
 	err = storageParams.MetadataChecker.CheckDefaultWorkloadStorage(cloudType, clusterMetadata.NominatedStorageClass)
-	if errors.IsNotFound(err) {
+	if errors.IsNotFound(err) && storageParams.WorkloadStorage == "" {
 		return fail(UnknownClusterError{CloudName: cloudType})
 	}
 	if storageParams.WorkloadStorage == "" && caas.IsNonPreferredStorageError(err) {
 		npse := err.(*caas.NonPreferredStorageError)
 		return fail(NoRecommendedStorageError{Message: err.Error(), ProviderName: npse.Name})
 	}
-	if err != nil && !caas.IsNonPreferredStorageError(err) {
+	maybeCustomStorage := err == nil || caas.IsNonPreferredStorageError(err) || errors.IsNotFound(err)
+	if err != nil && !maybeCustomStorage {
 		return fail(errors.Trace(err))
 	}
 
-	// If no storage class exists, we need to create one with the opinionated defaults.
+	// If no storage class exists, we need to create one with the opinionated defaults,
+	// or use an existing one.
 	var storageMsg string
-	if storageParams.WorkloadStorage != "" && caas.IsNonPreferredStorageError(err) {
-		preferredStorage := errors.Cause(err).(*caas.NonPreferredStorageError).PreferredStorage
+	if storageParams.WorkloadStorage != "" && maybeCustomStorage {
+		var (
```

Comment:
so does it mean if there is an existing sc named `--storage foo` found but which is not configured to Juju preferred provisioner, we will modify the foo to a Juju prefered config?
If so, we should probably not modify that.

Reply:
If there is an existing sc, EnsureStorageProvisioner() will exit early. So we don't modify it.
If --storage is used, we will try and use the sc the user has specified. If that doesn't exist, and we can't query the cluster to get a preferred storage type, we will exit with not found.

Reply:
```go
if storageParams.WorkloadStorage == ""{
	if errors.IsNotFound(err) {
		err = UnknownClusterError{CloudName: cloudType}
	}
	if caas.IsNonPreferredStorageError(err){
		npse := err.(*caas.NonPreferredStorageError)
		err = NoRecommendedStorageError{Message: err.Error(), ProviderName: npse.Name}
	}
}

if err != nil {
	return errors.Trace(err)
}

if storageParams.WorkloadStorage != "" {
	
}
```